### PR TITLE
Change password feature completed

### PR DESCRIPTION
**Pull request by beep101**
Reset forgotten password feature
External email service used

### CODE REVIEW

File: ui/src/components/newPassword.js [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        this.setState({[e.target.name]:e.target.value})
+    }
+    
+    render(){
+        return(
+            <div className="formContainer">
+
+                <div>
+                    <label className="inputLabel">Password</label><br/>
+                    <input className="inputFieldWide" id="passwd" name="password" type="password"  onChange={this.onChange} value={this.state.password}/><br/>
```
Comment (farukatlant) [inline ui/src/components/newPassword.js]: final break not necessary

File: ui/src/components/newPassword.js [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+                </div>
+                
+                <button className="buttonWide" onClick={this.newPassReq}>Change password</button>
+                {this.state.msg==""?null:<p className="warningMessage">{this.state.msg}</p>}
```
Comment (farukatlant) [inline ui/src/components/newPassword.js]: we have interesting improvement here with `this.state.msg && <p ...`

File: ui/src/apiConsumer/accountConsumer.js [hunk 3]
```diff
@@ -52,3 +52,26 @@ export function signup(userData,handler){
...
+            handler(true,"");
+        },
+        (error)=>{
+            handler(false,"");
+        }
+    );
+}
+
+export function newPassword(token,password,handler){
+    post("newPassword",{fpToken:token,password:password}).then(
```
Comment (farukatlant) [inline ui/src/apiConsumer/accountConsumer.js]: rename fp to forgotPassword everywhere, easier to understand

File: ui/src/apiConsumer/accountConsumer.js [hunk 4]
```diff
@@ -52,3 +52,26 @@ export function signup(userData,handler){
     );
 }
 
+export function forgotPassword(email,handler){
+    post("forgotPassword",{email:email}).then(
```
Comment (farukatlant) [inline ui/src/apiConsumer/accountConsumer.js]: we can just pass { email } here

File: src/main/java/com/example/demo/services/AccountService.java [hunk 5]
```diff
@@ -25,11 +31,17 @@
 	private UsersRepository usersRepo;
 	private IHashUtil hashUtil;
 	private IJwtUtil jwtUtil;
+	private JavaMailSender mailSender;
 	
-	public AccountService(IHashUtil hashUtil,IJwtUtil jwtUtil, UsersRepository userRepo) {
+	private String subject="Reset your Auction Purple password";
+	private String content="Follow link below to set new password for your Auction purple account";
+	private String link="http://localhost:8080/newPassword?token=";
```
Comment (farukatlant) [inline src/main/java/com/example/demo/services/AccountService.java]: don't use localhost:8000, will be very dificult when deployed to heroku to manage this

File: src/main/java/com/example/demo/services/AccountService.java [hunk 6]
```diff
@@ -75,9 +87,56 @@ public UserModel signUp(UserModel signup)throws InvalidDataException,ExistingUse
...
 
+	
 	@Override
-	public UserModel forgotPassword(UserModel forgotPassword) {
-		return null;
+	public UserModel forgotPassword(UserModel forgotPassword) throws NonExistentUserException {
+		if( forgotPassword.getEmail()==null||forgotPassword.getEmail().equals("")) {
+			throw new NonExistentUserException();
+		}
+		List<User> users=usersRepo.findByEmail(forgotPassword.getEmail());
```
Comment (farukatlant) [inline src/main/java/com/example/demo/services/AccountService.java]: make findByEmail return single user

File: src/main/java/com/example/demo/services/AccountService.java [hunk 7]
```diff
@@ -75,9 +87,56 @@ public UserModel signUp(UserModel signup)throws InvalidDataException,ExistingUse
...
+	@Override
+	public UserModel newPassword(UserModel newPassword) throws InvalidTokenException,InvalidDataException {
+		if(newPassword.getFpToken()==null || newPassword.getFpToken().equals("")) {
+			throw new InvalidTokenException();
+		}
+		if(newPassword.getPassword()==null || newPassword.getPassword().equals("")) {
+			throw new InvalidDataException();
+		}
+		
+		List<User> users=usersRepo.findByFpTokenAndFpTokenEndtimeAfter(newPassword.getFpToken(),new Timestamp(System.currentTimeMillis()));
```
Comment (farukatlant) [inline src/main/java/com/example/demo/services/AccountService.java]: similar, change to Optional<User> for return value

File: src/main/java/com/example/demo/entities/User.java [hunk 8]
```diff
@@ -34,7 +36,13 @@
 	
 	@Column(name="passwd")
 	private String passwd;
-
+	
+	@Column(name="fptoken")
+	private String fpToken;
+	
+	@Column(name="fptokenendtime")
```
Comment (farukatlant) [inline src/main/java/com/example/demo/entities/User.java]: rename to forgot_password_token_end_time

File: src/main/java/com/example/demo/services/AccountService.java [hunk 9]
```diff
@@ -75,9 +87,56 @@ public UserModel signUp(UserModel signup)throws InvalidDataException,ExistingUse
...
+		}
+		List<User> users=usersRepo.findByEmail(forgotPassword.getEmail());
+		if(users.size()==0) {
+			throw new NonExistentUserException();
+		}
+		
+		User user=users.get(0);
+		String token=UUID.randomUUID().toString().replace("-","");
+		user.setFpToken(token);
+		user.setFpTokenEndtime(new Timestamp(System.currentTimeMillis()+24*60*60*1000));
```
Comment (farukatlant) [inline src/main/java/com/example/demo/services/AccountService.java]: extract 24*60*60*1000 to constant with descriptive

File: src/test/java/com/example/demo/services/AccountServiceTests.java [hunk 10]
```diff
@@ -1,29 +1,41 @@
...
 import com.example.demo.exceptions.ExistingUserException;
 import com.example.demo.exceptions.NonExistentUserException;
 import com.example.demo.exceptions.InvalidDataException;
+import com.example.demo.exceptions.InvalidTokenException;
 import com.example.demo.models.UserModel;
 import com.example.demo.repositories.UsersRepository;
 import com.example.demo.utils.IHashUtil;
 import com.example.demo.utils.IJwtUtil;
 
+import ch.qos.logback.classic.pattern.ClassOfCallerConverter;
```
Comment (farukatlant) [inline src/test/java/com/example/demo/services/AccountServiceTests.java]: not used

File: ui/src/components/newPassword.js [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+            </div>
+        )
+    }
+
+    newPassReq=(e)=>{
+        let token=queryString.parse(this.props.location.search)['token'];
```
Comment (farukatlant) [inline ui/src/components/newPassword.js]: can be const

**Review by farukatlant — CHANGES_REQUESTED** [review]
Few improvements

File: ui/src/components/newPassword.js (hunk 2 above)
Comment (beep101) [reply ui/src/components/newPassword.js]: ???

Comment (farukatlant) [reply ui/src/components/newPassword.js]: this is not professional way to reply to comment

**Review by farukatlant — APPROVED** [review]

**Pull request merged**